src/emulator/tasksEmulator.ts

```diff
@@ -0,0 +1,285 @@
+
+export interface TasksEmulatorArgs {
+  port?: number;
+  host?: string;
+}
+
+export interface Task {
+  name: string;
+  // A timestamp in RFC3339 UTC "Zulu" format, with nanosecond resolution and up to nine fractional
+  // digits. Examples: "2014-10-02T15:01:23Z" and "2014-10-02T15:01:23.045123456Z".
+  scheduleTime?: string;
+  // A duration in seconds with up to nine fractional digits, terminated by 's'. Example: "3.5s".
+  dispatchDeadline?: string;
+  httpRequest: {
+    url: string;
+    oidcToken?: {
+      serviceAccountEmail: string;
+    };
+    body: any;
+    headers: { [key: string]: string };
+  };
+}
+
+export interface TaskQueueConfig {
+  retryConfig: RetryConfig;
+  rateLimits: RateLimits;
+  timeoutSeconds: number;
+  retry: boolean;
+  defaultUri: string;
+
+  // The configurations below this point do not currently have any effect on how the task queues are handled within the emulator
+  secrets?: string[]; // TODO(gburroughs): look into how we can handle this
+  region?: string | ResetValue;
+  memory?: string | ResetValue;
+  minInstances?: number | ResetValue;
+  maxInstances?: number | ResetValue;
+  concurrency?: number | ResetValue;
+  labels?: Record<string, string>;
+}
+
+export interface RetryConfig {
+  maxAttempts: number;
+  maxRetrySeconds: number | null;
+  maxBackoffSeconds: number;
+  maxDoublings: number;
+  minBackoffSeconds: number;
+}
+
+export interface RateLimits {
+  maxConcurrentDispatches: number;
+  maxDispatchesPerSecond: number;
+}
+
+type ResetValue = null;
+
+export class TaskQueueController {
+  static UPDATE_TIMEOUT = 500;
+  static LISTEN_TIMEOUT = 1000;
+  static TOKEN_REFRESH_INTERVAL = 1000;
+  queues: { [key: string]: TaskQueue } = {};
+  private listenId: NodeJS.Timeout | null;
+  private tokenRefillIds: NodeJS.Timeout[] = [];
+  private running = false;
+
+  constructor() {
+    this.listenId = null;
+  }
+
+  enqueue(key: string, task: Task): void {
+    if (!this.queues[key]) {
+      throw new Error("Queue does not exist");
+    }
+    this.queues[key].enqueue(task);
+  }
+
+  delete(key: string, taskId: string): void {
+    if (!this.queues[key]) {
+      throw new Error("Queue does not exist");
+    }
+    this.queues[key].delete(taskId);
+  }
+
+  createQueue(key: string, config: TaskQueueConfig): void {
+    const newQueue = new TaskQueue(key, config);
+    const intervalID = setInterval(
+      () => newQueue.refillTokens(),
+      TaskQueueController.TOKEN_REFRESH_INTERVAL,
+    );
+    this.tokenRefillIds.push(intervalID);
+    this.queues[key] = newQueue;
+  }
+
+  listen(): void {
+    let shouldUpdate = false;
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    for (const [_key, queue] of Object.entries(this.queues)) {
+      shouldUpdate = shouldUpdate || queue.isActive();
+    }
+    if (shouldUpdate) {
+      this.updateQueues();
+      this.listenId = setTimeout(() => this.listen(), TaskQueueController.UPDATE_TIMEOUT);
```

Comment:
From what I can tell, what I think your implementation seems to be doing (though I may be misunderstanding) is batching the task processing to occur every 500 ms. Alternatively, we can consider using asynchronous callbacks to continuously pull tasks from the queue (while respecting rate limits). In any case, we should add a comment to this `listen()` function to explain how you've decided to implement task processing behavior, since as far as I can tell it isn't immediately obvious.

Also, why 500 ms as the timeout?

Reply:
Oops, I set the interval pretty high just for testing reasons, but the idea was to have it much lower (1-10ms)

It isn't meant so much as a batch process (although it may handle multiple tasks at a single time if the tasks are getting dispatched fast enough) but rather a continuous loop to rapidly process tasks when there are tasks to be handled, and then when there are no more tasks to handle, it should wait longer (1s intervals) between checking if any more tasks have been queued.  

I'll update the interval + leave a comment explaining the logic if you think this makes sense.